File: app/src/main/java/name/monwf/customiuizer/subs/SortableList.java
package name.monwf.customiuizer.subs;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Toast;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.UUID;

import name.monwf.customiuizer.R;
import name.monwf.customiuizer.SubFragment;
import name.monwf.customiuizer.utils.AppHelper;
import name.monwf.customiuizer.utils.Helpers;
import name.monwf.customiuizer.utils.PreferenceAdapter;
import name.monwf.customiuizer.utils.SortableListView;

@SuppressWarnings("ConstantConditions")
public class SortableList extends SubFragment {

	String key;
	String titleResId;
	boolean activities;
	SortableListView listView;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		this.padded = false;
		super.onCreate(savedInstanceState);
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		toolbarMenu = true;
		super.onActivityCreated(savedInstanceState);

		Bundle args = getArguments();
		key = args.getString("key");
		titleResId = args.getString("titleResId");
		activities = args.getBoolean("activities", false);

		if (getView() == null) return;

		listView = getView().findViewById(android.R.id.list);

		if (!activities) try {
			Field ssField = SortableListView.class.getDeclaredField("mSnapshotShadow");
			ssField.setAccessible(true);
			int lightShadow = getResources().getIdentifier("dynamic_listview_dragging_item_shadow_light", "drawable", "miui");
			int darkShadow = getResources().getIdentifier("dynamic_listview_dragging_item_shadow_dark", "drawable", "miui");
			ssField.set(listView, getResources().getDrawable(Helpers.isNightMode(getContext()) ? darkShadow : lightShadow, getContext().getTheme()));
		} catch (Throwable e) {
			e.printStackTrace();
		}

		listView.setAdapter(new PreferenceAdapter(getContext(), key, activities));
		if (activities)	listView.setOnOrderChangedListener(null); else
		listView.setOnOrderChangedListener(new SortableListView.OnOrderChangedListener() {
			@Override
			public void OnOrderChanged(int oldPos, int newPos) {
				if (oldPos == newPos) return;
				String itemStr = AppHelper.getStringOfAppPrefs(key, "");
				if (itemStr.isEmpty()) return;
				String[] itemArr = itemStr.trim().split("\\|");
				ArrayList<String> itemList = new ArrayList<String>(Arrays.asList(itemArr));
				String uuid = itemList.get(oldPos);
				itemList.remove(oldPos);
				itemList.add(newPos, uuid);
				AppHelper.appPrefs.edit().putString(key, TextUtils.join("|", itemList)).apply();
				((PreferenceAdapter)listView.getAdapter()).updateItems();
				((PreferenceAdapter)listView.getAdapter()).notifyDataSetChanged();
			}
		});
		if (!activities) {
			listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
				@Override
				public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
					String uuid = ((PreferenceAdapter)parent.getAdapter()).getItem(position);
					Bundle args = new Bundle();
					args.putString("key", key + "_" + uuid);
					args.putInt("actions", MultiAction.Actions.LOCKSCREEN.ordinal());
					openSubFragment(new MultiAction(), args, Helpers.SettingsType.Edit, Helpers.ActionBarType.Edit, titleResId, R.layout.prefs_multiaction);
				}
			});
		}
		listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
			@Override
			public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
				deleteItem(position);

//				PopupMenu popup = new PopupMenu(getContext(), view);
//				popup.inflate(R.menu.menu_itemoptions);
//				popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
//					@Override
//					public boolean onMenuItemClick(MenuItem menuItem) {
//						switch (menuItem.getItemId()) {
//							case R.id.changeicon:
//								float density = getContext().getResources().getDisplayMetrics().density;
//								GuidePopup iconsPopup = new GuidePopup(getContext());
//								GridView grid = new GridView(getContext());
//								grid.setNumColumns(4);
//								grid.setAdapter(new IconGridAdapter(getContext()));
//								grid.setBackgroundColor(Color.rgb(108, 108, 111));
//								grid.setPadding(Math.round(7 * density), Math.round(5 * density), 0, Math.round(5 * density));
//								grid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
//									@Override
//									public void onItemClick(AdapterView<?> parent2, View view2, int position2, long id2) {
//										String uuid = ((PreferenceAdapter)parent.getAdapter()).getItem(position);
//										AppHelper.prefs.edit().putString(key + "_" + uuid + "_icon", ((IconGridAdapter)parent2.getAdapter()).getItem(position2)).apply();
//										iconsPopup.dismiss(true);
//										((PreferenceAdapter)parent.getAdapter()).notifyDataSetChanged();
//									}
//								});
//								iconsPopup.setContentView(grid);
//								iconsPopup.setArrowMode(parent.getChildAt(0) == view ? 1 : 0);
//								iconsPopup.setWidth(Math.round(200 * density));
//								iconsPopup.setContentWidth(Math.round(200 * density));
//								iconsPopup.showAsDropDown(view);
//								break;
//							case R.id.deleteitem:
//								PreferenceAdapter adapter = (PreferenceAdapter)listView.getAdapter();
//								String uuid = adapter.getItem(position);
//								String items = AppHelper.prefs.getString(key, "");
//								AppHelper.prefs.edit().putString(key, items == null || items.isEmpty() ? "" : items.replace(uuid, "").replace("||", "|").replaceAll("^\\|", "").replaceAll("\\|$", "")).apply();
//								adapter.updateItems();
//								adapter.notifyDataSetChanged();
//								invalidateOptionsMenu();
//								break;
//						}
//						return true;
//					}
//				});
//				popup.show();
				return true;
			}
		});
	}

	private String createNewUUID() {
		return UUID.randomUUID().toString().replaceAll("-", "").toLowerCase();
	}

	private void createNewItem(String uuid) {
		String items = AppHelper.getStringOfAppPrefs(key, "");
		AppHelper.appPrefs.edit().putString(key, items.isEmpty() ? uuid : items + "|" + uuid).apply();
		PreferenceAdapter adapter = (PreferenceAdapter)listView.getAdapter();
		adapter.updateItems();
		adapter.notifyDataSetChanged();
	}

	private void deleteItem(int position) {
		PreferenceAdapter adapter = (PreferenceAdapter)listView.getAdapter();
		String items = AppHelper.getStringOfAppPrefs(key, "");
		AppHelper.appPrefs.edit().putString(key, items.isEmpty() ? "" : items.replace(adapter.getItem(position), "").replace("||", "|").replaceAll("^\\|", "").replaceAll("\\|$", "")).apply();
		adapter.updateItems();
		adapter.notifyDataSetChanged();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item.getItemId() == R.id.deleteitem) {
			Toast.makeText(getContext(), R.string.delete_item_info, Toast.LENGTH_SHORT).show();
			return true;
		} else if (item.getItemId() == R.id.additem) {
			if (activities) {
				Bundle args = new Bundle();
				args.putBoolean("activity", true);
				args.putString("key", key);
				AppSelector activitySelect = new AppSelector();
				activitySelect.setTargetFragment(SortableList.this, 2);
				openSubFragment(activitySelect, args, Helpers.SettingsType.Edit, Helpers.ActionBarType.HomeUp, R.string.select_app, R.layout.prefs_app_selector);
			} else {
				//if (listView.getChildCount() >= 10) return true;
				createNewItem(createNewUUID());
			}
			return true;
		}
		return super.onOptionsItemSelected(item);
	}

	public void onPrepareOptionsMenu(Menu menu) {}

	@Override
	public void onActivityResult(int requestCode, int resultCode, final Intent data) {
		if (resultCode == Activity.RESULT_OK) {
			if (requestCode == 2) {
				String activityValue = data.getStringExtra("activity");
				int activityUser = data.getIntExtra("user", 0);
				if (activityValue == null || activityUser < 0) return;

				String uuid = createNewUUID();
				AppHelper.appPrefs.edit().putInt(key + "_" + uuid + "_action", 20).putString(key + "_" + uuid + "_activity", activityValue).putInt(key + "_" + uuid + "_activity_user", activityUser).apply();
				createNewItem(uuid);
			}
		}
		super.onActivityResult(requestCode, resultCode, data);
	}

}

File: app/src/main/java/tv/withaibuild/customiuizer/subs/SortableList.kt
package tv.withaibuild.customiuizer.subs

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import tv.withaibuild.customiuizer.R
import tv.withaibuild.customiuizer.SubFragment
import tv.withaibuild.customiuizer.utils.AppHelper
import tv.withaibuild.customiuizer.utils.Helpers
import tv.withaibuild.customiuizer.utils.PreferenceAdapter
import tv.withaibuild.customiuizer.utils.SortableListView
import java.util.UUID
import java.util.Locale

class SortableList : SubFragment() {

    private var key: String? = null
    private var titleResId: String? = null
    private var activities = false
    private var listView: SortableListView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        this.padded = false
        super.onCreate(savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        toolbarMenu = true
        super.onActivityCreated(savedInstanceState)

        val args = arguments
        key = args?.getString("key")
        titleResId = args?.getString("titleResId")
        activities = args?.getBoolean("activities", false) ?: false

        val v = view ?: return
        val ctx = context ?: return
        val mKey = key ?: return

        listView = v.findViewById(android.R.id.list)

        if (!activities) try {
            val ssField = SortableListView::class.java.getDeclaredField("mSnapshotShadow")
            ssField.isAccessible = true
            val lightShadow = ctx.resources.getIdentifier("dynamic_listview_dragging_item_shadow_light", "drawable", "miui")
            val darkShadow = ctx.resources.getIdentifier("dynamic_listview_dragging_item_shadow_dark", "drawable", "miui")
            val drawable = ctx.resources.getDrawable(if (Helpers.isNightMode(ctx)) darkShadow else lightShadow, ctx.theme)
            ssField.set(listView, drawable)
        } catch (e: Throwable) {
            e.printStackTrace()
        }

        listView?.adapter = PreferenceAdapter(ctx, mKey, activities)
        if (activities) {
            listView?.setOnOrderChangedListener(null)
        } else {
            listView?.setOnOrderChangedListener(object : SortableListView.OnOrderChangedListener {
                override fun OnOrderChanged(oldPos: Int, newPos: Int) {
                    if (oldPos == newPos) return
                    val itemStr = AppHelper.getStringOfAppPrefs(mKey, "") ?: return
                    if (itemStr.isEmpty()) return
                    val itemList = itemStr.trim().split("|").toMutableList()
                    val uuid = itemList.removeAt(oldPos)
                    itemList.add(newPos, uuid)
                    AppHelper.appPrefs.edit().putString(mKey, itemList.joinToString("|")).apply()
                    (listView?.adapter as? PreferenceAdapter)?.run {
                        updateItems()
                        notifyDataSetChanged()
                    }
                }
            })
        }

        if (!activities) {
            listView?.setOnItemClickListener { parent, _, position, _ ->
                val uuid = (parent.adapter as? PreferenceAdapter)?.getItem(position) ?: return@setOnItemClickListener
                val itemArgs = Bundle().apply {
                    putString("key", mKey + "_" + uuid)
                    putInt("actions", MultiAction.Actions.LOCKSCREEN.ordinal)
                }
                openSubFragment(MultiAction(), itemArgs, AppHelper.SettingsType.Edit, AppHelper.ActionBarType.Edit, titleResId, R.layout.prefs_multiaction)
            }
        }

        listView?.setOnItemLongClickListener { _, _, position, _ ->
            deleteItem(position)
            true
        }
    }

    private fun createNewUUID(): String =
        UUID.randomUUID().toString().replace("-", "").lowercase(Locale.ROOT)

    private fun createNewItem(uuid: String) {
        val mKey = key ?: return
        val items = AppHelper.getStringOfAppPrefs(mKey, "") ?: ""
        AppHelper.appPrefs.edit().putString(mKey, if (items.isEmpty()) uuid else "$items|$uuid").apply()
        (listView?.adapter as? PreferenceAdapter)?.run {
            updateItems()
            notifyDataSetChanged()
        }
    }

    private fun deleteItem(position: Int) {
        val mKey = key ?: return
        val adapter = listView?.adapter as? PreferenceAdapter ?: return
        val items = AppHelper.getStringOfAppPrefs(mKey, "") ?: ""
        if (items.isEmpty()) return
        val itemList = items.split("|").toMutableList().apply { removeAt(position) }
        AppHelper.appPrefs.edit().putString(mKey, itemList.joinToString("|")).apply()
        adapter.updateItems()
        adapter.notifyDataSetChanged()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val mKey = key
        return when (item.itemId) {
            R.id.deleteitem -> {
                Toast.makeText(context, R.string.delete_item_info, Toast.LENGTH_SHORT).show()
                true
            }
            R.id.additem -> {
                if (activities) {
                    val args = Bundle().apply {
                        putBoolean("activity", true)
                        putString("key", mKey)
                    }
                    val activitySelect = AppSelector().apply { setTargetFragment(this@SortableList, 2) }
                    openSubFragment(activitySelect, args, AppHelper.SettingsType.Edit, AppHelper.ActionBarType.HomeUp, R.string.select_app, R.layout.prefs_app_selector)
                } else {
                    createNewItem(createNewUUID())
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {}

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK && requestCode == 2 && data != null) {
            val mKey = key ?: return
            val activityValue = data.getStringExtra("activity") ?: return
            val activityUser = data.getIntExtra("user", 0)
            if (activityUser < 0) return

            val uuid = createNewUUID()
            AppHelper.appPrefs.edit()
                .putInt(mKey + "_" + uuid + "_action", 20)
                .putString(mKey + "_" + uuid + "_activity", activityValue)
                .putInt(mKey + "_" + uuid + "_activity_user", activityUser)
                .apply()
            createNewItem(uuid)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }
}
